src/eventloop.jl

```diff
@@ -1,73 +1,122 @@
+(note: in IJulia the shell socket is called `requests`).
+"""
+function waitloop(kernel)
+    control_msgs = Channel{Msg}(32) do ch
+        task_local_storage(:IJulia_task, "control msgs receive task")
+        while isopen(kernel.control[])
             try
-                msg = recv_ipython(socket, kernel)
+                msg::Msg = recv_ipython(kernel.control[], kernel)
+                put!(ch, msg)
             catch e
-                if isa(e, EOFError)
-                    # The socket was closed
+                if IJulia._shutting_down[] || isa(e, EOFError)
+                    # an EOFError is because of a closed socket
                     return
                 else
                     rethrow()
                 end
             end
+            yield()
+        end
+    end
 
+    iopub_msgs = Channel{Msg}(32)
+    request_msgs = Channel{Msg}(32) do ch
+        task_local_storage(:IJulia_task, "request msgs receive task")
+        while isopen(kernel.requests[])
             try
-                send_status("busy", kernel, msg)
-                invokelatest(get(handlers, msg.header["msg_type"], unknown_request), socket, kernel, msg)
+                msg::Msg = recv_ipython(kernel.requests[], kernel)
+                if haskey(iopub_handlers,  msg.header["msg_type"])
+                    put!(iopub_msgs, msg)
+                else
+                    put!(ch, msg)
+                end
             catch e
-                if e isa InterruptException && IJulia._shutting_down[]
-                    # If we're shutting down, just return immediately
+                if IJulia._shutting_down[] || isa(e, EOFError)
+                    close(iopub_msgs) # otherwise iopubs_msg would remain open, but with no producer anymore
+                    # an EOFError is because of a closed socket
                     return
-                elseif !isa(e, InterruptException)
-                    # Try to keep going if we get an exception, but
-                    # send the exception traceback to the front-ends.
-                    # (Ignore SIGINT since this may just be a user-requested
-                    #  kernel interruption to interrupt long calculations.)
-                    content = error_content(e, msg="KERNEL EXCEPTION")
-                    map(s -> println(orig_stderr[], s), content["traceback"])
-                    send_ipython(kernel.publish[], kernel, msg_pub(kernel.execute_msg, "error", content))
+                else
+                    rethrow()
                 end
-            finally
-                flush_all()
-                send_status("idle", kernel, msg)
             end
-        end
-    catch e
-        if IJulia._shutting_down[]
-            return
-        end
-
-        # the Jupyter manager may send us a SIGINT if the user
-        # chooses to interrupt the kernel; don't crash on this
-        if isa(e, InterruptException)
-            eventloop(socket, kernel)
```

Comment:
I can't think of a reason for this to be recursive instead of the outer loop as I reorganized it too?